Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Kms/KmsXmlEncryptor.cs

```diff
@@ -50,12 +53,17 @@ public EncryptedXmlInfo Encrypt(XElement plaintextElement)
                 plainTextData = ByteString.FromStream(stream);
             }
 
+            if (plainTextData.Length > MaximumEncryptSize)
+            {
+                throw new ArgumentException($"Cannot encrypt data of size {plainTextData.Length}. Maximum supported size: {MaximumEncryptSize}");
+            }
```

Comment:
If the user is unlucky enough to exceed this limit, what should they do?  Suddenly, they're exceeding a limit they probably didn't know existed (I know it's documented, but chances are they never read noticed), and the don't have a clue how to cope with it.

I'd rather solve this problem for users so they never hit it, by using envelope encryption.

Reply:
That would involve adding somewhere else to store something, as the body of the envelope. We'd end up having to generate a new key, locally encrypt the body using that key, then encrypt the key material with KMS. As I said, ASP.NET Core is already *using* envelope encryption, effectively (to the best of my understanding). Adding an extra level of indirection would make this more complicated for no benefit, IMO.

Would it help if I were to take a look at what ends up being encrypted at the moment, when using the data protection API for the purpose of anti-forgery validation? I can also see whether there's any guidance on expectations of the data protection API in terms of payloads. Finally, I'm happy to see whether the Azure encryptor has a similar restriction - if it does (and it's the same or smaller) would that be good enough?

Reply:
I take it back - I think I may have misunderstood what the Azure Key Value code does before. I need to look into this further...

Reply:
Acknowledged.  The only benefit I seek from envelope encryption is the ability to encrypt > 64k.

Reply:
Please see the third commit, which significantly changes everything - we can now encrypt arbitrary sizes of data (modulo it ending up in an XML element as base64...)